library/alloc/src/collections/vec_deque/mod.rs

```diff
@@ -2372,72 +2088,61 @@ impl<T, A: Allocator> VecDeque<T, A> {
         }
 
-        let buf = self.buf.ptr();
+        let &mut Self { head, len, .. } = self;
+        let ptr = self.ptr();
         let cap = self.cap();
-        let len = self.len();
 
-        let free = self.tail - self.head;
-        let tail_len = cap - self.tail;
+        let free = cap - len;
+        let head_len = cap - head;
+        let tail = len - head_len;
+        let tail_len = tail;
 
-        if free >= tail_len {
-            // there is enough free space to copy the tail in one go,
-            // this means that we first shift the head backwards, and then
-            // copy the tail to the correct position.
+        if free >= head_len {
+            // there is enough free space to copy the head in one go,
+            // this means that we first shift the tail backwards, and then
+            // copy the head to the correct position.
             //
             // from: DEFGH....ABC
             // to:   ABCDEFGH....
             unsafe {
-                ptr::copy(buf, buf.add(tail_len), self.head);
+                self.copy(0, head_len, tail_len);
                 // ...DEFGH.ABC
-                ptr::copy_nonoverlapping(buf.add(self.tail), buf, tail_len);
+                self.copy_nonoverlapping(head, 0, head_len);
                 // ABCDEFGH....
-
-                self.tail = 0;
-                self.head = len;
             }
-        } else if free > self.head {
-            // FIXME: We currently do not consider ....ABCDEFGH
-            // to be contiguous because `head` would be `0` in this
-            // case. While we probably want to change this it
-            // isn't trivial as a few places expect `is_contiguous`
-            // to mean that we can just slice using `buf[tail..head]`.
 
-            // there is enough free space to copy the head in one go,
-            // this means that we first shift the tail forwards, and then
-            // copy the head to the correct position.
+            self.head = 0;
+        } else if free >= tail_len {
+            // there is enough free space to copy the tail in one go,
+            // this means that we first shift the head forwards, and then
+            // copy the tail to the correct position.
             //
             // from: FGH....ABCDE
             // to:   ...ABCDEFGH.
             unsafe {
-                ptr::copy(buf.add(self.tail), buf.add(self.head), tail_len);
+                self.copy(head, tail, head_len);
                 // FGHABCDE....
-                ptr::copy_nonoverlapping(buf, buf.add(self.head + tail_len), self.head);
+                self.copy_nonoverlapping(0, tail + head_len, tail_len);
                 // ...ABCDEFGH.
-
-                self.tail = self.head;
-                self.head = self.wrap_add(self.tail, len);
             }
+
+            self.head = tail;
         } else {
             // free is smaller than both head and tail,
```

Comment:
I also don't like this algorithm that much, but I didn't wanna wrap my head around it for this PR, so I just reused the existing implementation.

Reply:
If anyone wants to tackle this in the future, I think it's actually pretty easy since what you need already exists on slices.
1. memmove to go from `EFGHI...ABCD` -> `EFGHIABCD...`.
2. then `buf[..len].rotate_right(capacity - head)`, thanks to https://doc.rust-lang.org/std/primitive.slice.html#method.rotate_right, to get `ABCDEFGHI...`

Or the same thing at the right end instead if that's less data movement, of course.

(The two special cases before this `else` branch are worth keeping too, though, for when we have simpler ways.)

Reply:
Oh nice, I can try to implement that then. I guess technically rotating the entire buffer is slightly inefficient, because it does extra work to also move the unused parts of the buffer, but because of the above if statements it means that the unused space is at most 1/3 of the entire buffer I think.